subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r51/ProjectConfigurationProgressEventCrossVersionSpec.groovy

```diff
@@ -349,6 +340,17 @@ class ProjectConfigurationProgressEventCrossVersionSpec extends ToolingApiSpecif
         result.totalConfigurationTime < Duration.ofMillis(2 * sleepDurationMillis)
     }
 
+    def simulateWork(long durationMillis) {
+        """
+            def start = org.gradle.internal.time.Time.currentTimeMillis()
+            Thread.sleep($durationMillis)
+            def elapsed
+            while ((elapsed = org.gradle.internal.time.Time.currentTimeMillis() - start) < $durationMillis) {
+                Thread.sleep($durationMillis - elapsed)
+            }
```

Comment:
This doesn't seem very right.

Suppose `start = 100`, then after `Thread.sleep(100)`, when `elapsed = org.gradle.internal.time.Time.currentTimeMillis() - start` is calculated, current time might be `200`, `300`, `1000`, or more, i.e. If the thread happens to wait for a long time between `Thread.sleep($durationMillis)` and `(elapsed = org.gradle.internal.time.Time.currentTimeMillis() - start)`, `elapsed` might be very large, then `$durationMillis - elapsed` is negative. `Thread.sleep` says

```
* @throws  IllegalArgumentException
     *          if the value of {@code millis} is negative
```

Can't we just replace original code to

```
long start = org.gradle.internal.time.Time.currentTimeMillis()
while (org.gradle.internal.time.Time.currentTimeMillis() - start < $sleepMillis) {
    Thread.sleep($sleepMillis)
}
```

if `System.currentTimeMillis()` implementation is not accurate on Windows?

Reply:
`durationMillis - elapsed` is never negative because the `while` loop is only entered/repeated if `elapsed < durationMillis`.

Reply:
Ah... sorry I was wrong. Then no questions.